fix(forms): allow control state as reset argument for `FormGroup`

bougwal opened the pull request:
fix(forms): Allow ControlState as reset arguments for `FormGroup`/`FormRecord`

This change also decorelate the `reset` type argument from `TValue` by adding a 3rd generic parameter to `AbstractControl`. 
This improves the typings overall. 

This PR closes https://github.com/angular/angular/issues/55577.

JeanMeche commented on packages/forms/src/model/form_group.ts:
```diff
@@ -32,7 +33,7 @@ import {
 export type ɵFormGroupValue<T extends {[K in keyof T]?: AbstractControl<any>}> = ɵTypedOrUntyped<
   T,
-  Partial<{[K in keyof T]: ɵValue<T[K]>}>,
+  Partial<{[K in keyof T]: ɵValue<T[K]> | FormControlState<ɵValue<T[K]>>}>,
```
Changing the type of the value with a way to big breaking change, I think we'll need to find another solution.

Beside that I don't think this is even correct, `FormControlState` is only used when creating controls, the type probably shouldn't appear in the value signature.

JeanMeche replied:
Actually you're on the right track with. 
There 2 steps we need to take here
1.  Create a new type something like out of your suggestion 

```
export type ɵFormGroupArgumentValue<T extends {[K in keyof T]?: AbstractControl<any>}> = ɵTypedOrUntyped<
  T,
  Partial<{[K in keyof T]: ɵValue<T[K]> | FormControlState<ɵValue<T[K]>>}>,
  {[key: string]: any}
>;
```

2. Replace the type in the `reset` signature : 

```
  override reset(
    value: ɵTypedOrUntyped<
      TControl,
      ɵFormGroupArgumentValue<TControl>,
      any
    > = {},
    options: {onlySelf?: boolean; emitEvent?: boolean} = {},
  ): void
```

bougwal replied:
Guess what ? I was on the right track right before raising the PR. I opted to a similar approach before raising my PR ❤️, then I decided to avoid it duo to the fact that my changes may look over-done creating a new type. So I went ahead with inline change - I agree with you that from a design standpoint the existing approach is sub-optimal as it might be breaking and there might be no need for exposing that change publicly as you said. 
____
Let me Switch to the suggestion ! Thanks Matthieu

JeanMeche reviewed (changes requested):
See my suggestions.

JeanMeche commented on packages/forms/test/typed_integration_spec.ts:
```diff
@@ -167,7 +167,6 @@ describe('Typed Class', () => {
       });
       {
         type ValueType = Partial<{c: string; d: number}>;
-        let t: ValueType = c.value;
```
Can you revert those changes ? Thank you.

bougwal replied:
Done !

bougwal committed:
fix(forms): Allow ControlState as reset arguments for `FormGroup`/`FormRecord`

This change also decorelate the `reset` type argument from `TValue` by adding a 3rd generic parameter to `AbstractControl`.
This improves the typings overall.

JeanMeche commented:
Thank for the changes Walid !  
We'll do some checks against the Google codebase and we'll let you know if we can go forward with the change.

dylhunn commented:
We're investigating an issue illustrated by the following snippet:
```ts
function valueChangesWrapper<T>(control: AbstractControl<T>) {
  return control.valueChanges;
}
const fg = new FormGroup({foo: new FormControl('')});
fg.valueChanges;
const a = valueChangesWrapper(fg);
```

Notice that the inferred type of `valueChangesWrapper` inappropriately includes `FormControlState`. We think this is due to type unification of `FormGroup.reset` with `AbstractControl.reset`.

JeanMeche commented:
I've allowed myself to push the changes to fix the issues that were brought up by testing the fix against Google's code base. 

Basically what happens is that changing the reset() signature had implicationg on the `TValue` generic of `AbstractControl`.

GuillaumeCz commented:
Hello ! 

Any update on this PR ? ;)

kirjs commented:
I'm not super familiar with this area, 
Do I understand correctly, that this behavior is something that is implemented, but wasn't supported on the  type level?
How would it disambiguate between an object that has a `disabled` prop, and `value` props, and setting up state?

JeanMeche commented:
[TGP is "green"](https://fusion2.corp.google.com/presubmit/753946468/OCL:753946468:BASE:753957630:1746187075931:a467b15).

JeanMeche commented:
@kirjs to answer your question, you'd have to do the same as you would declare such control: the state like structure must be defined in a state. 

    const fc = new FormControl({ value: {value: 'test', disabled: true}, disabled: true });
    fc.reset({ value: {value: 'xx', disabled: false}, disabled: true })
    console.log(fc.value); // {value: 'xx', disabled: false}

JeanMeche reviewed (approved):
We wrote the changes with @dylhunn.

kirjs reviewed (approved):
Reviewed-for: public-api

thePunderWoman reviewed (approved):
reviewed-for: public-api

thePunderWoman commented:
This PR was merged into the repository by commit 610bebfce98d879677244b2ef08b24886891ca76.

The changes were merged into the following branches: main
